### PR TITLE
TLSv1.3: Allow ciphersuite to change on resumption

### PR DESCRIPTION
**Pull request by mattcaswell**
In <=TLSv1.2 during resumption you must always select the same ciphersuite as the original session. However in TLSv1.3 it is only necessary for a compatible ciphersuite to be used (one where the hash is the same as the original session). Previously on the client side we were bailing out if the server selected a different ciphersuite. On the server side we now select the ciphersuite first (TLSv1.3 only) prior to resumption (consistent with the recommendations in the draft-20). This could also result in a different ciphersuite being selected but resumption still taking place.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by tmshort** [comment]
Should there be an option to control this behavior?

**Comment by kaduk** [comment]
> Should there be an option to control this behavior?

What behavior would you want such an option to change?

**Comment by tmshort** [comment]
> What behavior would you want such an option to change?

To force the cipher to be the same; potentially for badly behaving clients. But such a case is likely to be hard to identify. Thinking about SSL_OP_CIPHER_SERVER_PREFERENCE, but not sure how that might impact the behavior.

With significant behavior changes, it's usually desirable to have a way to disable/revert the behavior. But TLSv1.3 is new; so not sure it matters as this is merely one aspect.

**Comment by mattcaswell** [comment]
> With significant behavior changes, it's usually desirable to have a way to disable/revert the behavior. But TLSv1.3 is new; so not sure it matters as this is merely one aspect.

The behaviour change is only for TLSv1.3 which as you say is new and is so different in so many respects that I don't think it matters. Clients will have to be able to handle a change in ciphersuite because servers might do it no matter what settings we put in. On the server side this brings us into line with an explicit recommendation in the spec, so it should not be surprising behaviour for clients to have to deal with. I don't think we should be putting in options to work around bugs that we don't yet know exist! If buggy clients appear and we think it is a sufficiently large problem then we could decide to add an option at that point - but I think the added complexity is not worth it now.

**Comment by mattcaswell** [comment]
Rebased to resolve conflicts, and added a commit to fix a travis failure. Ping @openssl/commiters?

**Comment by kaduk** [comment]
Interesting that b9b8f98ac74844e9b431fe83cda96fef9e55c2ed reverts some of the additions from 5f4c48ab1f91dd4878bdf47533f7f528769db5aa, but I didn't look at how hard it would be to switch the order in which they appear.

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -1850,7 +1881,7 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
      * Given s->session->ciphers and SSL_get_ciphers, we must pick a cipher
      */
 
-    if (!s->hit || s->hello_retry_request) {
+    if (!s->hit || SSL_IS_TLS13(s)) {
```
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: I'm having trouble seeing how this chunk relates to the stated change.  Is it unrelated cleanup?

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -1956,9 +1987,9 @@ WORK_STATE tls_post_process_client_hello(SSL *s, WORK_STATE wst)
         wst = WORK_MORE_B;
     }
     if (wst == WORK_MORE_B) {
-        if (!s->hit || s->hello_retry_request) {
+        if (!s->hit || SSL_IS_TLS13(s)) {
```
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: Likewise for this hello_retry_request=>SSL_IS_TLS13() conversion; is it unrelated cleanup?

File: ssl/statem/statem_srvr.c [hunk 3]
```diff
@@ -1566,6 +1566,68 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
...
+                SSLerr(SSL_F_TLS_EARLY_POST_PROCESS_CLIENT_HELLO,
+                       SSL_R_INAPPROPRIATE_FALLBACK);
+                al = SSL_AD_INAPPROPRIATE_FALLBACK;
+                goto err;
+            }
+        }
+    }
+
+    /* For TLSv1.3 we must select the ciphersuite *before* session resumption */
+    if (SSL_IS_TLS13(s)) {
```
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: A little unfortunate that we cannot just move *all* ciphersuite selection prior to resumption (for code clearnness), but that would lead to a regression for some cases (failing to resume when previously we would), I think.

File: test/sslapitest.c [hunk 4]
```diff
@@ -1797,8 +1798,130 @@ static int test_early_data_tls1_2(int idx)
...
+                                       &cctx, cert, privkey))
+            || !TEST_true(SSL_CTX_set_cipher_list(cctx,
+                                                  "TLS13-AES-128-GCM-SHA256"))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl,
+                                          &clientssl, NULL, NULL))
+            || !TEST_true(create_ssl_connection(serverssl, clientssl,
+                                                SSL_ERROR_NONE)))
+        goto end;
+
+    clntsess = SSL_get1_session(clientssl);
```
Comment (kaduk) [inline test/sslapitest.c]: Aren't we supposed to use a new_session_cb to make sure we get a resumable session instead of just a placeholder (i.e., before the NewSessionTicket is received)?

File: test/sslapitest.c [hunk 5]
```diff
@@ -1797,8 +1798,130 @@ static int test_early_data_tls1_2(int idx)
...
+    serverssl = clientssl = NULL;
+
+    if (!TEST_true(SSL_CTX_set_cipher_list(cctx,
+                   "TLS13-AES-128-GCM-SHA256:TLS13-AES-256-GCM-SHA384"))
+            || !TEST_true(SSL_CTX_set_cipher_list(sctx,
+                                                  "TLS13-AES-256-GCM-SHA384"))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL))
+            || !TEST_true(SSL_set_session(clientssl, clntsess))
+            || !TEST_false(create_ssl_connection(serverssl, clientssl,
```
Comment (kaduk) [inline test/sslapitest.c]: Should this stanza get a comment that it's stopping mid-handshake to be able to futz with things?
(I missed the SSL_ERROR_WANT_READ the first time around, assuming it was SSL_ERROR_NONE like ~everything else.  Amusing, given I introduced that functionality...)

File: test/sslapitest.c [hunk 6]
```diff
@@ -1797,8 +1798,130 @@ static int test_early_data_tls1_2(int idx)
...
+                                                SSL_ERROR_WANT_READ)))
+        goto end;
+
+    /* Trick the client into thinking this session is for a different digest */
+    clntsess->cipher = aes_128_gcm_sha256;
```
Comment (kaduk) [inline test/sslapitest.c]: One might argue that we should get a fresh session object for this attempt, given discussion on the tlswg list about aiming for single-use tickets.  But of course we can always change this test when we implement that, so this should be considered a side note.

File: ssl/s3_lib.c [hunk 7]
```diff
@@ -3714,8 +3714,10 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
         allow = srvr;
     }
 
-    tls1_set_cert_validity(s);
-    ssl_set_masks(s);
+    if (!SSL_IS_TLS13(s)) {
+        tls1_set_cert_validity(s);
```
Comment (mattcaswell) [inline ssl/s3_lib.c]: tls1_set_cert_validity() calls tls1_check_chain() repeatedly for the various cert types. A lot of the checks in tls1_check_chain() only occur if suitb_flags is set, or strict_mode is set. These checks are only relevant to suiteb which is only well defined for TLSv1.2 (not TLSv1.3).

The only check that is actually relevant in this context to non-suiteb is the call to tls1_check_cert_param(). This only checks that the sig-algs and curve are appropriate to the certificate. In TLSv1.3 we do this instead in tls_choose_sigalg().

File: ssl/statem/statem_srvr.c (hunk 3 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: I seriously considered doing this but came to the same conclusion. :-(

File: ssl/statem/statem_srvr.c (hunk 1 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: No, it is relevant. The point is that we are now going to select a ciphersuite in TLSv1.3 whether or not we have resumed. In <= TLSv1.2 we only did this if we did not resume (because a resumption implies we've already selected the ciphersuite).

File: ssl/statem/statem_srvr.c (hunk 2 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: As above.

File: test/sslapitest.c (hunk 4 above)
Comment (mattcaswell) [inline test/sslapitest.c]: If we were a normal application then, yes. But we are a test so we can make some assumptions about how our server is going to behave. We know that an OpenSSL server always sends a NewSessionTicket immediately after the end of the handshake. create_ssl_connection() does not complete until it has received that ticket, so at this point we know it is safe to call SSL_get1_session() (obviously if we later change the server behaviour this won't work any more - but the test will fail in that case so we can fix it then).

File: test/sslapitest.c (hunk 5 above)
Comment (mattcaswell) [inline test/sslapitest.c]: Done

File: test/sslapitest.c (hunk 6 above)
Comment (mattcaswell) [inline test/sslapitest.c]: The session has only been used once at this stage in any case.

**Comment by mattcaswell** [comment]
I added a new commit to add extra comments to sslapitest.c as per above, and also rebased to resolve a conflict.

**Comment by kaduk** [comment]
Travis claims

> The command "$make update; git diff --quiet" exited with 1.

but only on the one builder.  Can we explain it?

**Comment by mattcaswell** [comment]
Yes. It's unrelated to this PR. Rather it is because of a problem with #3686. @levitte is already on the case.

File: ssl/s3_lib.c (hunk 7 above)
Comment (kaduk) [inline ssl/s3_lib.c]: The sigalgs was the only bit I found that I was pretty sure was relevant.
So, tls_choose_sigalg() covers that and this is safe, thanks.

Though, in looking at how tmp.peer_sigalgs gets set, I found myself looking at tls_parse_ctos_sig_algs() and wondering whether it is correct to only call tls1_save_sigalgs() if not resuming, as for TLS 1.3 these extensions should be renegotiated on each handshake, IIRC.

File: ssl/statem/statem_srvr.c [hunk 8]
```diff
@@ -1566,6 +1566,68 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
...
+        const SSL_CIPHER *cipher =
+            ssl3_choose_cipher(s, ciphers, SSL_get_ciphers(s));
+
+        if (cipher == NULL) {
+            SSLerr(SSL_F_TLS_EARLY_POST_PROCESS_CLIENT_HELLO,
+                   SSL_R_NO_SHARED_CIPHER);
+            al = SSL_AD_HANDSHAKE_FAILURE;
+            goto err;
+        }
+        if (s->hello_retry_request && s->s3->tmp.new_cipher != NULL
```
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: If we send a HRR, we must have picked a cipher already.
So, wouldn't s->s3->tmp.new_cipher == NULL be an error in its own right?
(That is, I don't think we need to check it here.)

File: ssl/statem/statem_srvr.c [hunk 9]
```diff
@@ -1716,7 +1743,11 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
         }
     }
 
-    if (!s->hit && s->version >= TLS1_VERSION && s->ext.session_secret_cb) {
+    if (!s->hit
+            && s->version >= TLS1_VERSION
+            && !SSL_IS_TLS13(s)
+            && !SSL_IS_DTLS(s)
+            && s->ext.session_secret_cb) {
```
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: Conveniently (?), SSL_set_session_secret_cb() is undocumented, so there is no documentation to update to indicate that it cannot be used with TLS 1.3.
(I will not hold this work hostage to writing such a man page.)

File: ssl/statem/statem_srvr.c (hunk 1 above)
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: Ah, got it.
(I think maybe I was reading the || as an && the first time, though I'm not sure anymore.)

File: ssl/statem/statem_srvr.c (hunk 2 above)
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: This one seems to just be about not taking the "else" branch, as everything inside this branch is either excluded if SSL_IS_TLS13() or requires !s->hit.
But sure, it makes sense now -- thanks for the explanation.

File: test/sslapitest.c (hunk 6 above)
Comment (kaduk) [inline test/sslapitest.c]: Right.  I must have written this before I saw the SSL_ERROR_WANT_READ handshake-stop-point.

File: ssl/statem/statem_srvr.c (hunk 8 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: I tweaked it like this:

````
diff --git a/ssl/statem/statem_srvr.c b/ssl/statem/statem_srvr.c
index 0f55d2652d..6f57816810 100644
--- a/ssl/statem/statem_srvr.c
+++ b/ssl/statem/statem_srvr.c
@@ -1615,8 +1615,9 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
             al = SSL_AD_HANDSHAKE_FAILURE;
             goto err;
         }
-        if (s->hello_retry_request && s->s3->tmp.new_cipher != NULL
-                && s->s3->tmp.new_cipher->id != cipher->id) {
+        if (s->hello_retry_request
+                && (s->s3->tmp.new_cipher == NULL
+                    || s->s3->tmp.new_cipher->id != cipher->id)) {
             /*
              * A previous HRR picked a different ciphersuite to the one we
              * just selected. Something must have changed.
````

So I still check its not NULL before I attempt to deref it (for safety), but now treat it as an error if it is NULL.

File: ssl/statem/statem_srvr.c (hunk 9 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: Yes, it should be documented at some point. FYI, this primarily exists for EAP-FAST support I believe (which is a fairly obscure corner of libssl).

**Comment by mattcaswell** [comment]
Pushed. Thanks.

File: ssl/statem/statem_srvr.c (hunk 3 above)
Comment (kaduk) [inline ssl/statem/statem_srvr.c]: On the other hand, if we assume that a given client's cipherlist is not going to change much over time, isn't the regression a one-time thing when the server upgrades its openssl version?  That is, a session issued by a new server should still be resumable at the new server, if the client is sending the same cipher list in the connection attempting to reuse the session and the connection where the session was issued.
This is somewhat relevant for (HTTP/2) protocol compliance, as we can use ALPN to negotiate whether h2 is used, but h2 has a (TLS 1.2) cipher suite blacklist -- peers are permitted to generate an INADEQUATE_SECURITY connection error when a blacklisted cipher is used.  So, ideally an application alpn_select_cb would have access to the cipher suite to be used, but at present for non-TLS-1.3 connections that's not the case.

Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: Its still a significant behaviour changed which probably can't be done before 1.2.0

Comment (kaduk) [inline ssl/statem/statem_srvr.c]: Is it really a bigger change than (e.g.) our great extensions refactoring?  To take a concrete example, in 1.1.0 the alpn_select_cb ran after cipher negotiation, thanks to some custom hackery introduced in commit 817cd0d52f0462039d1fe60462150be7f59d2002.  But commit 02f0274e8c0596dcf7e2d104250232a42c650b96 moved this processing out of ssl_check_clienthello_tlsext_late() and into an extension finalizer.  (I guess it was only considering ordering with respect to SNI processing and not cipher negotiation.)

That's only one example, of course, but I find it pretty likely that there are other noticeable behavior changes.

Comment (davidben) [inline ssl/statem/statem_srvr.c]: We did this in BoringSSL. It's a little trickier than merely removing this check.

The events that end up happening in the old order are:

1. Decide on resumption.
2. Parse extensions (call SNI callback).
3. Call `cert_cb`.
4. Select cipher suite.

The goal is to order (4) before (1), but merely moving (4) doesn't quite work. (4) needs to happen after (2) and (3) because, until the certificate is fixed, you cannot select a TLS 1.2 cipher. You don't know whether to filter out the RSA or ECDSA cipher suite. One of the goals in redoing TLS 1.3 cipher negotiation was precisely to avoid this interaction. The consumer may also have different cipher configurations per SNI or whatever, which also imposes this constraint. Whether you wish to support the latter is still open since TLS 1.3 didn't exist in 1.1.0.

In BoringSSL, we reordered to (2), (3), (4), (1) with this patch series:
* https://boringssl.googlesource.com/boringssl.git/+/4eb95ccfd64d46fad23f36238690594d02518467 (ignore the ALPN note in the commit message; I'd forgotten about [this change](https://boringssl.googlesource.com/boringssl/+/9ef31f01af6c59efe65fe0c4ab3be7316046ccca))
* https://boringssl.googlesource.com/boringssl.git/+/34202b93b6195e9c55402a6bc2653956d0cd26d8
* https://boringssl.googlesource.com/boringssl.git/+/f01f42a2ce161657c6413ef71d7c04262c7abd56

But it had some user-visible effects (I documented the ones I could think of in the commit messages) and we had to patch one consumer. They previously used `SSL_get_session` + `session->tlsext_hostname` which we switched to `SSL_get_servername`. But the end result is much simpler.

Hope this is helpful!
